### ISSUE TITLE
Incorrect scaling on Steam Deck screen

### ISSUE THREAD

**Issue by Jacalz** [fyne-io/fyne#4896]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

Copying over from https://github.com/Jacalz/rymdport/issues/159, Fyne applications are entirely unusable on Steam Deck (maybe even other small screens?). This is really problematic because anyone using a Steam Deck and then installing something like Rymdport or Supersonic on Flathub will be very annoyed because the apps have stopped working.

### How to reproduce

1. Start any Fyne application on Steam Deck.
2. Notice that scaling makes it impossible to use the app.

### Screenshots

![image](https://github.com/fyne-io/fyne/assets/25466657/41b05c8f-8794-43d2-9513-28ab3aee7d15)
![image](https://github.com/fyne-io/fyne/assets/25466657/1a79cc8e-e9b8-468e-86de-ea37c6e9e1cb)

### Example code

Use Fyne demo or install Rymdport or Supersonic from Flathub.

### Fyne version

v2.4.5

### Go compiler version

1.22.3

### Operating system and version

SteamOS

### Additional Information

See https://github.com/Jacalz/rymdport/issues/159 for more information.

**Comment by andydotxyz** [comment]
I don't think this can be universally true as I have seen many apps running on the steam deck. Is it related to user configuration or is this a recent problem?

**Comment by andydotxyz** [comment]
Any more information at all? Can the user run fyne_demo or otherwise collect scale data from the device?

**Comment by Jacalz** [comment]
I haven't quite had time to research this more. Maybe we can bring in @daliz-bazalgette can chime in here?

**Comment by dalitz-bazalgette** [comment]
I did a bit more testing and this scaling issue seems to only happen on X11 (which is what SteamOS uses).

I installed Fedora 39 (because it uses KDE 5.X like SteamOS) on my Steam Deck and found that Rymdport and Supersonic scale correctly in wayland but in X11 they exhibit the issue as they do in SteamOS.

**Comment by andydotxyz** [comment]
There must be something more specific going on because X11 normally works really well.
Can you share what hardware / configuration you are testing the X11 with?

**Comment by dalitz-bazalgette** [comment]
I'm using the Valve Steam Deck. I did a fresh install of SteamOS and Fedora to make sure it wasn't my configuration that caused it.

Perhaps it's a combination of X11 and the 800x1280 display that causes it.

**Comment by Jacalz** [comment]
For what it is worth, I saw that I have an old but similar report from another user having bad scaling in Rymdport:
![image](https://github.com/fyne-io/fyne/assets/25466657/a9ea9382-59a9-47f8-8244-abe743245c13)

**Comment by Jacalz** [comment]
@dalitz-bazalgette Thanks for the testing. Does Fedora 40 with what I presume is KDE Plasma 6.1 get the same behaviour? I also found it strange that X11 shows issues but not Wayland. As the app is rendering through XWayland, it is usually as good or slightly worse than a pure X11 session.

**Comment by andydotxyz** [comment]
> Perhaps it's a combination of X11 and the 800x1280 display that causes it.

The resolution should not matter - as long as the screen size is reported correctly... my guess is the RandR API has some incorrect data.

Is it possible to run the fyne_demo app on the device and choose "Advanced" so we can see what information is being detected?

**Comment by andydotxyz** [comment]
I have tracked down the source of the issue - the screen is reported (to GLFW at least) as 6x6cm. Instead of being 15.6cm wide. So the scaling is off - it thinks the UI should be 4.5x normal.
If I "fix" that with a small hack in our driver then the scale is determined as 1.7 which sounds about right and gives a good resolution of text:
![Screenshot_20240708_224020](https://github.com/fyne-io/fyne/assets/294436/3f843717-2680-4a84-9a17-e22e0fa33a2e)

I'm not sure how we can fix this at the GLFW layer though - if the hardware reports it wrong...

**Comment by Jacalz** [comment]
Nice work looking into it. Is there any other way to look up scaling and screen sizes on X11? I mean, many of the other applications seem to be working as far as I understand it.

**Comment by andydotxyz** [comment]
> Is there any other way to look up scaling and screen sizes on X11?

That is what we are doing, it is being reported incorrectly (as is a known issue with some hardware - though I'm not sure if this is the XWayland or underlying hardware/OS at fault).

> I mean, many of the other applications seem to be working as far as I understand it.

Yes. They are doing no attempt at scaling to fit the screen, they are simply 1:1 on pixels, leading to very small text because of the underlying hardware DPI. (see the image above - the desktop text is smaller than Fyne based on a scale of 1 whereas my hacked screen size fix works it out to be 1.7 for the right font size.

**Comment by andydotxyz** [comment]
The PR has a workaround where the scaling isn't correct according to our aims, but is in-keeping with the tiny text of other apps on the system

**Comment by andydotxyz** [comment]
OMG I am loving the simplicity of just dropping some Fyne game onto SteamDeck :)

![IMG_4952](https://github.com/fyne-io/fyne/assets/294436/efa15ec2-74a4-4cdb-a56b-9afab32bfe2b)

**Comment by andydotxyz** [comment]
Fixed on `develop` for v2.5.0
